### ISSUE TITLE
A  problem about generating BaggingClassifier with wrapped classifier  

### ISSUE THREAD

**Issue by Jerry2990** [adriangb/sklearn_keras_wrap#4]
Hi, adrian:
Really appreciate your work. i read your some comments in other issues.
However, i have a question. When generating a BaggingClassifier with classifier wrapped from keras:
```
classifier = sklearn_keras_wrap.wrappers.KerasClassifier(build_fn=model, batch_size=64, epochs=1)
classifier.fit(train_x, train_y, batch_size=64, epochs=1)
ensemble_clf = BaggingClassifier(classifier, n_estimators=n_estimators)
ensemble_clf.fit(train_x, train_y)  #error
```` 
An error will occur:  
```
rv = reductor(4)
TypeError: can't pickle _thread.RLock objects
```
It is mentioned many times. Do you have any targeted suggestions for me about modifying bagging.py?
Thank you very much. Look forward your reply.

**Comment by adriangb** [comment]
Could you show me what `model` is in your test, or post a fully reproducible example?

**Comment by Jerry2990** [comment]
Thank you for your reply. This is a VERY useful work.
I solved my question. There is nothing wrong with your code.
The reason is that tensors in estimator generated by keras can't be deepcopied correctly with function **clone** in sklearn/base.py. I added a copy method based on your wrapper code. But, this still involves changes to code in ensemble/base.py(two lines of code).
```
        if hasattr(self.base_estimator_, 'build_fn'):
            estimator = self.base_estimator_.duplicate()
        else:
            estimator = clone(self.base_estimator_)
```
 In order not to pollute the original code I had to re-implement them under my path. I haven't thought of any other way yet.

**Comment by adriangb** [comment]
This should be working without having to modify anything in sklearn. I'll try it tonight with an arbitrary Keras model, but could please post a full example? Specifically, what is your `model`? Is it a prebuilt model? A function that returns a model?

**Comment by Jerry2990** [comment]
> This should be working without having to modify anything in sklearn. I'll try it tonight with an arbitrary Keras model, but could please post a full example? Specifically, what is your model? Is it a prebuilt model? A function that returns a model?

No problem！ I'll post a simple example later.

**Comment by Jerry2990** [comment]
This is a simple example for test.
`
import numpy as np
from math import floor

from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Dense, Flatten, Reshape, Conv1D, MaxPooling1D
from tensorflow import keras

from sklearn.ensemble import (BaggingClassifier, AdaBoostClassifier, GradientBoostingClassifier)


from sklearn_keras_wrap.wrappers import KerasClassifier as wrapper

def build_model(data_shape,  num_classes):
    model = Sequential()
    model.add(Reshape((data_shape, 1), name='input', input_shape=(data_shape,)))
    model.add(Conv1D(16, 3, padding='same', activation='relu', name='conv1'))
    model.add(MaxPooling1D(pool_size=2, name='pool1'))

    model.add(Flatten(name='flat'))
    model.add(Dense(32, activation='relu', name='dense1'))
    # model.add(Dropout(0.5))
    model.add(Dense(num_classes, activation='softmax', name='output'))

    opt = keras.optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0, amsgrad=False)
    # Let's train the model using RMSprop
    model.compile(loss='sparse_categorical_crossentropy',
                  optimizer=opt,
                  metrics=['accuracy'])
    return model

if __name__ == "__main__":
    train_x, train_y = np.random.rand(1000, 64), np.random.random_integers(0,10,size=(1000,1))

    model = build_model(train_x.shape[1], np.unique(train_y).shape[0])

    sub_classifier = wrapper(build_fn=model, batch_size=64, epochs=1)
    sub_classifier.fit(train_x, train_y, batch_size=64, epochs=1)

    ensemble_clf = BaggingClassifier(sub_classifier, n_estimators=5, max_samples=floor(0.8 * train_x.shape[0]))
    ensemble_clf.fit(train_x, train_y)
`
There may be some problems when cloning estimator. Thank you for your help.

**Comment by adriangb** [comment]
Thank you. I was able to confirm this is a problem with the wrappers. Let me try and fix it and I'll get back to you. Thank you again for reporting this!

**Comment by adriangb** [comment]
@Jerry2990 I put a fix up at #5. Would you mind reviewing the changes? Any feedback is appreciated.

**Comment by Jerry2990** [comment]
Really appreciate you. 
The latest version solves the problem very well. I will continue following this work.
Thanks again!
